src/Cryptol/TypeCheck/Subst/Type.hs

```diff
@@ -0,0 +1,275 @@
+    classify x =
+      case fst x of
+        TVFree i _ _ _ -> Left (i, x)
+        TVBound tp -> Right (tpUnique tp, x)
+
+-- | Makes a substitution out of a list.
+-- WARNING: We do not validate the list in any way, so the caller should
+-- ensure that we end up with a valid (e.g., idempotent) substitution.
+listParamSubst :: [(TParam, Type)] -> Subst
+listParamSubst xs
+  | null xs   = emptySubst
+  | otherwise = S { suFreeMap = IntMap.empty
+                  , suBoundMap = IntMap.fromList bounds
+                  , suDefaulting = False }
+  where
+    bounds = [ (tpUnique tp, (TVBound tp, t)) | (tp, t) <- xs ]
+
+isEmptySubst :: Subst -> Bool
+isEmptySubst su = IntMap.null (suFreeMap su) && IntMap.null (suBoundMap su)
+
+-- Returns the empty set if this is a defaulting substitution
+substBinds :: Subst -> Set TVar
+substBinds su
+  | suDefaulting su = Set.empty
+  | otherwise       = Set.fromList (map fst (assocsSubst su))
+
+substToList :: Subst -> [(TVar, Type)]
+substToList s
+  | suDefaulting s = panic "substToList" ["Defaulting substitution."]
+  | otherwise = assocsSubst s
+
+assocsSubst :: Subst -> [(TVar, Type)]
+assocsSubst s = frees ++ bounds
+  where
+    frees = IntMap.elems (suFreeMap s)
+    bounds = IntMap.elems (suBoundMap s)
+
+instance PP (WithNames Subst) where
+  ppPrec _ (WithNames s mp)
+    | null els  = text "(empty substitution)"
+    | otherwise = text "Substitution:" $$ nest 2 (vcat (map pp1 els))
+    where pp1 (x,t) = ppWithNames mp x <+> text "=" <+> ppWithNames mp t
+          els       = assocsSubst s
+
+instance PP Subst where
+  ppPrec n = ppWithNamesPrec IntMap.empty n
+
+-- | Apply a substitution.  Returns `Nothing` if nothing changed.
+-- The @Prop -> Prop@ function is how to simplify the result if it is a
+-- predicate.
+apSubstMaybe' :: (Prop -> Prop) -> Subst -> Type -> Maybe Type
+apSubstMaybe' simpProp su ty =
+  case ty of
+    TCon t ts ->
+      do ss <- anyJust (apSubstMaybe' simpProp su) ts
+         case t of
+           TF _ -> Just $! Simp.tCon t ss
+           PC _ -> Just $! simpProp (TCon t ss)
+           _    -> Just (TCon t ss)
+
+    TUser f ts t ->
+      do (ts1, t1) <- anyJust2 (anyJust (apSubstMaybe' simpProp su))
+                               (apSubstMaybe' simpProp su)
+                               (ts, t)
+         Just (TUser f ts1 t1)
+
+    TRec fs -> TRec `fmap` (anyJust (apSubstMaybe' simpProp su) fs)
+
+    {- We apply the substitution to nominal types as well, because it might
+    contain module parameters, which need to be substituted when
+    instantiating a functor. -}
+    TNominal nt ts ->
+      uncurry TNominal <$> anyJust2 (applySubstToNominalType' simpProp su)
+                                    (anyJust (apSubstMaybe' simpProp su))
+                                    (nt,ts)
+
+    TVar x -> applySubstToVar' (apSubstMaybe' simpProp) su x
+
+-- | Apply a substitution without simplifying the result when it is a predicate.
```

Comment:
Maybe add a note here on why we need this.  As far as I understand, we use it while instantiating nominal types in the class solver, and we don't want to simplify stuff because we are in  the process of solving, so we'd get in a loop.  Is that right?